src/_pytest/assertion/rewrite.py

```diff
@@ -777,24 +793,46 @@ def visit_Assert(self, assert_):
                     top_condition, module_path=self.module_path, lineno=assert_.lineno
                 )
             )
-        # Create failure message.
-        body = self.on_failure
         negation = ast.UnaryOp(ast.Not(), top_condition)
-        self.statements.append(ast.If(negation, body, []))
+        msg = self.pop_format_context(ast.Str(explanation))
         if assert_.msg:
             assertmsg = self.helper("_format_assertmsg", assert_.msg)
-            explanation = "\n>assert " + explanation
+            gluestr = "\n>assert "
         else:
             assertmsg = ast.Str("")
-            explanation = "assert " + explanation
-        template = ast.BinOp(assertmsg, ast.Add(), ast.Str(explanation))
-        msg = self.pop_format_context(template)
-        fmt = self.helper("_format_explanation", msg)
+            gluestr = "assert "
+        err_explanation = ast.BinOp(ast.Str(gluestr), ast.Add(), msg)
+        err_msg = ast.BinOp(assertmsg, ast.Add(), err_explanation)
         err_name = ast.Name("AssertionError", ast.Load())
+        fmt = self.helper("_format_explanation", err_msg)
+        fmt_pass = self.helper("_format_explanation", msg)
         exc = ast.Call(err_name, [fmt], [])
-        raise_ = ast.Raise(exc, None)
+        if sys.version_info[0] >= 3:
+            raise_ = ast.Raise(exc, None)
+        else:
+            raise_ = ast.Raise(exc, None, None)
+        # Call to hook when passes
+        orig = astor.to_source(assert_.test).rstrip("\n").lstrip("(").rstrip(")")
```

Comment:
I really would feel better if this assertion rewriting part could be made opt-in through an option.

Reply:
You think the risk lies when calling the assertion pass hook or simply on the assertion rewriting changes?

In other words, should this switch just omit the `if` that calls the hook (but keeping the structural changes compared to before), or the switch should actually revert the entire rewriting to be exactly the same as before?

Reply:
> You think the risk lies when calling the assertion pass hook or simply on the assertion rewriting changes?

The latter. 👍 

> In other words, should this switch just omit the if that calls the hook (but keeping the structural changes compared to before), or the switch should actually revert the entire rewriting to be exactly the same as before?

You got it, to revert the entire rewriting to be the same as before.

My fear is that, because there's a ton of code in the wild, things can go wrong in the assertion code itself, blowing up for users which don't even use the feature. This might be the kind of "major breakage" which I wouldn't like to see in the 5.0 release, hence I would prefer we skip the entire rewriting if the option is not set.

How hard you think it would be to include this switch? If it would complicate the code considerably we can skip it, but if it is simple to implement it, we should do it. 😁 

If you end up including the option, make sure to mention it in the docstring for the hook, so it is clear that users should set it for the hook to work.

As for the name of the option, I suggest `enable_pytest_assertion_pass_hook` defaulting to `False`, but feel to suggest something else.

Reply:
Thanks! I added the option, although it was a little tricky to get it down to the assertion rewriter as the options are not yet parsed so we get it with `config.getvalue` it seems. You can take a look.

Also, one thing that might be problematic is that pytest rewrites and caches even site-packages libraries, so they should have the pyc files removed as well.